api/src/main/java/org/pmiops/workbench/cohortbuilder/TemporalQueryBuilder.java

```diff
@@ -0,0 +1,124 @@
+import static org.pmiops.workbench.cohortbuilder.querybuilder.util.Validation.from;
+
+/**
+ * TemporalQueryBuilder is an object that builds {@link QueryJobConfiguration}
+ * for BigQuery for the Temporal criteria groups.
+ */
+@Service
+public class TemporalQueryBuilder {
+
+  private static final String UNION_TEMPLATE = "union all\n";
+  private static final String SAME_ENC =
+    "temp1.person_id = temp2.person_id and temp1.visit_concept_id = temp2.visit_concept_id\n";
+  private static final String X_DAYS_BEFORE =
+    "temp1.person_id = temp2.person_id and temp1.entry_date <= DATE_SUB(temp2.entry_date, INTERVAL ${timeValue} DAY)\n";
+  private static final String X_DAYS_AFTER =
+    "temp1.person_id = temp2.person_id and temp1." +
+      "entry_date >= DATE_ADD(temp2.entry_date, INTERVAL ${timeValue} DAY)\n";
+  private static final String WITHIN_X_DAYS_OF =
+    "temp1.person_id = temp2.person_id and temp1.entry_date between " +
+      "DATE_SUB(temp2.entry_date, INTERVAL ${timeValue} DAY) and DATE_ADD(temp2.entry_date, INTERVAL ${timeValue} DAY)\n";
+  private static final String TEMPORAL_EXIST_TEMPLATE =
+    "select count(distinct temp1.person_id)\n" +
+      "from (${query1}) temp1\n" +
+      "where exists (select 1\n" +
+      "from (${query2}) temp2\n" +
+      "where (${conditions}))\n";
+  private static final String TEMPORAL_JOIN_TEMPLATE =
+    "select count(distinct temp1.person_id)\n" +
+      "from (${query1}) temp1\n" +
+      "join (select person_id, visit_concept_id, entry_date\n" +
+      "from (${query2})\n" +
+      ") temp2 on (${conditions})\n";
+
+  public String buildQuery(Map<String, QueryParameterValue> params,
+                           SearchGroup includeGroup) {
+    validateSearchGroup(includeGroup);
+    List<String> temporalQueryParts1 = new ArrayList<>();
+    List<String> temporalQueryParts2 = new ArrayList<>();
+    ListMultimap<Integer, SearchGroupItem> temporalGroups = getTemporalGroups(includeGroup);
+    for (Integer key : temporalGroups.keySet()) {
+      List<SearchGroupItem> tempGroups = temporalGroups.get(key);
+      //key of zero indicates belonging to the first temporal group
+      //key of one indicates belonging to the second temporal group
+      boolean isFirstGroup = key == 0;
+      for (SearchGroupItem tempGroup : tempGroups) {
+        String query = QueryBuilderFactory
+          .getQueryBuilder(FactoryKey.getType(tempGroup.getType()))
+          .buildQuery(params, tempGroup,
+            isFirstGroup ? includeGroup.getMention() : TemporalMention.ANY_MENTION.name());
+        if (isFirstGroup) {
+          temporalQueryParts1.add(query);
+        } else {
+          temporalQueryParts2.add(query);
+        }
+      }
+    }
+    String conditions = SAME_ENC;
+    String parameterName = "p" + params.size();
+    params.put(parameterName, QueryParameterValue.int64(includeGroup.getTimeValue()));
+    if (includeGroup.getTime().equals(TemporalTime.WITHIN_X_DAYS_OF.name())) {
```

Comment:
The type of this field should actually be a TemporalTime enum type, rather than an untyped string.